compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirInlineClassDeclarationChecker.kt

```diff
@@ -0,0 +1,135 @@
+        }
+
+        val parameter = primaryConstructor.valueParameters.singleOrNull() ?: run {
+            reporter.reportOn(primaryConstructor.source, FirErrors.INLINE_CLASS_CONSTRUCTOR_WRONG_PARAMETERS_SIZE, context)
+            return
+        }
+
+        when {
+            parameter.isNotFinalReadOnly(declaration) ->
+                reporter.reportOn(parameter.source, FirErrors.INLINE_CLASS_CONSTRUCTOR_NOT_FINAL_READ_ONLY_PARAMETER, context)
+
+            parameter.returnTypeRef.isInapplicableParameterType() ->
+                reporter.reportOn(
+                    parameter.returnTypeRef.source,
+                    FirErrors.INLINE_CLASS_HAS_INAPPLICABLE_PARAMETER_TYPE,
+                    parameter.returnTypeRef.coneType,
+                    context
+                )
+
+            parameter.returnTypeRef.coneType.isRecursiveInlineClassType(context.session) ->
+                reporter.reportOn(parameter.returnTypeRef.source, FirErrors.INLINE_CLASS_CANNOT_BE_RECURSIVE, context)
+        }
+    }
+
+    private fun FirDeclaration.isSecondaryConstructorWithBody() =
+        this is FirConstructor && !isPrimary && body != null
+
+    private fun FirValueParameter.isNotFinalReadOnly(declaration: FirRegularClass): Boolean {
+        val associatedProperty = declaration.getFirstPrimaryConstructorProperty() ?: return true
+
+        val modifierList = with(FirModifierList) { source.getModifierList() }
+        val isOpen = modifierList?.modifiers?.any { it.token == KtTokens.OPEN_KEYWORD } == true
+
+        return isVararg || !associatedProperty.isVal || isOpen
+    }
+
+    private fun FirTypeRef.isInapplicableParameterType() =
+        isUnit || isNothing || coneType is ConeTypeParameterType || coneType.isGenericArrayOfTypeParameter()
+
+    private fun ConeKotlinType.isGenericArrayOfTypeParameter(): Boolean {
+        if (this.typeArguments.firstOrNull() is ConeStarProjection || !isPotentiallyArray())
+            return false
+
+        val arrayElementType = arrayElementType()?.type ?: return false
+        return arrayElementType is ConeTypeParameterType ||
+                arrayElementType.isGenericArrayOfTypeParameter()
+    }
+
+    private fun ConeKotlinType.isRecursiveInlineClassType(session: FirSession) =
+        isRecursiveInlineClassType(hashSetOf(), session)
+
+    private fun ConeKotlinType.isRecursiveInlineClassType(visited: HashSet<ConeKotlinType>, session: FirSession): Boolean {
+        if (!visited.add(this)) return true
+
+        val asRegularClass = this.toRegularClass(session) ?: return false
+
+        return asRegularClass.isInlineOrValueClass() &&
+                asRegularClass.getRealPrimaryConstructor()
+                    ?.valueParameters
+                    ?.firstOrNull()
+                    ?.returnTypeRef
+                    ?.coneType
+                    ?.isRecursiveInlineClassType(visited, session) == true
+    }
+
+    private fun FirTypeRef.implementedByDelegation() =
```

Comment:
I suggest here another way which is a bit more complex but does not require tree structure manipulation. Class with delegated super types always contains `FirField` declaration (one per delegated super type), which has `isSynthetic = true`, and its `(returnTypeRef as FirResolvedTypeRef).delegatedTypeRef` has the same source as super type itself, so it can be used for reporting the relevant error.